<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>A crash course in Bayesian nonparametric inference with inferno</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">A crash course in Bayesian nonparametric
inference with inferno</h1>



<div id="before-we-start" class="section level2">
<h2>Before we start</h2>
<p>This vignette gives an introduction and guide to the kinds of
<em>Bayesian nonparametric inference</em> that can be done with
<strong><em>inferno</em></strong>, by means of a concrete example. It
also has the purpose to clarify, by means of examples, the terminology
used in the <strong><em>inferno</em></strong> package.</p>
<p>It is of course impossible to summarize this branch of probability
theory and of statistics in a couple of sections; you’re invited to
learn more for instance from the <a href="https://pglpm.github.io/ADA511">ADA 511</a> course, or from texts
such as Bernardo &amp; Smith’s <a href="https://doi.org/10.1002/9780470316870"><em>Bayesian Theory</em>
(2000)</a> (especially §§ 4.3, 4.4, 4.6), or Jaynes’s <a href="https://doi.org/10.1017/CBO9780511790423"><em>Probability
Theory</em> (2003)</a> (esp. ch. 9); take also a look at Lindley &amp;
Novick <a href="https://doi.org/10.1214/aos/1176345331"><em>The role of
exchangeability in inference</em> (1981)</a>.</p>
<p>Two extremely important warnings before you follow the example:</p>
<ul>
<li><p>Some terms may sound familiar to you, but keep in mind that they
may have quite different meanings from what you’re used to. This is
especially true of the terms in <em>italics</em>. When you see a term in
<strong>boldface</strong>, try to understand its meaning from the way
it’s used, rather than assuming the meaning familiar to you.
Unfortunately, different methods and different disciplines use the same
terms in different ways.</p>
<p>In particular keep in mind that the term <strong>probability</strong>
does <em>not</em> mean <strong>frequency</strong>. ‘Probability’ means
‘plausibility’ or ‘degree of belief’; it is a quantification of a
researcher’s uncertainty about some possible fact. There is a connection
between probability and frequency, but they are not the same. The
distinction is important, because one of the main problems in
statistical inference is that <em>we are uncertain about the frequency
of something</em> – the occurrence of a disease, symptom, or other
characteristics – and our goal is to quantify and reduce that
uncertainty as much as possible.</p></li>
<li><p>The example that follows is, like every example, very specific.
But try to see the more general, abstract picture behind it, and to see
how its general methodology could be applied to research that is more
interesting to you. From time to time we shall draw analogies with other
examples that may look very different and yet use essentially the same
methodology.</p></li>
</ul>
</div>
<div id="penguins" class="section level2">
<h2>Penguins</h2>
<p>We are researchers interested in penguins; specifically the
<strong>population</strong> living in the Antarctic islands
<em>Biscoe</em>, <em>Dream</em>, <em>Torgersen</em>. Our research
questions are not yet precisely defined, but some initial questions of
interest are the following. The penguins can be of three species: <a href="https://www.bas.ac.uk/about/antarctica/wildlife/penguins/adelie-penguin/"><em>Adélie</em></a>,
<a href="https://www.bas.ac.uk/about/antarctica/wildlife/penguins/chinstrap-penguin/"><em>Chinstrap</em></a>,
<a href="https://www.bas.ac.uk/about/antarctica/wildlife/penguins/gentoo-penguin/"><em>Gentoo</em></a>;
we’d like to know:</p>
<dl>
<dt>Q1</dt>
<dd>
Do the three species differ, <em>statistically</em>, in some physical or
geographical characteristic such as sex or island of origin? We say
“statistically” because, for instance, there surely are females and
males of every species, but it’s possible that for one of the species
the ratio of females to males is higher than for another species.
</dd>
<dt>Q2</dt>
<dd>
Is there a physical or geographical characteristic that allows us to
make a good guess about a penguin’s species? We say “guess” because, for
instance, from knowing which island a penguin comes from we can’t be
100% sure about that penguin’s species.
</dd>
</dl>
<p>In order to approach these and other future questions, we decide
which set of penguin characteristics we should consider and observe.
These characteristics are called the <strong>variates</strong> of the
population. This set can be extended or reduced later on. Guided by
previous studies, or by some hypotheses we are entertaining, we choose
the following variates, each denoted by a short
<code>codeword</code>:</p>
<ul>
<li>Species (<code>species</code>)</li>
<li>Island (<code>island</code>)</li>
<li>Bill (<code>bill_len</code>): <a href="https://allisonhorst.github.io/palmerpenguins/reference/figures/culmen_depth.png">the
length of a penguin’s “beak”</a> in millimetres</li>
<li>Bill depth (<code>bill_dep</code>): <a href="https://allisonhorst.github.io/palmerpenguins/reference/figures/culmen_depth.png">the
depth of a penguin’s “beak”</a> in millimetres</li>
<li>Flipper length (<code>flipper_len</code>): the length of a penguin’s
“wing” in millimetres</li>
<li>Body mass (<code>body_mass</code>) in grammes</li>
<li>Sex (<code>sex</code>)</li>
</ul>
<p>Later on we shall include an additional variate:</p>
<ul>
<li>Study year (<code>year</code>): the year a penguin was observed and
measured</li>
</ul>
<p>because there could be time trends in the values we observe. For
instance, a particular species might have more females than males during
one year, and vice versa during another year.</p>
<div id="population" class="section level4">
<h4>“Population”?</h4>
<p>The question of time-trends leads us to more general considerations
and questions about statistical research, which unfortunately are often
forgotten.</p>
<p>To start with: this “population” which we want to study, how is it
defined exactly? What counts as a member of the population, and what
doesn’t? For example, a seal clearly doesn’t count, because it isn’t a
penguin. A penguin from the Galápagos island doesn’t count either,
because it isn’t from one of the three Antarctic islands we specified.
But does a penguin who was alive in the year 1830 count? what about one
who will live in those islands in the year 2100? Does a penguin with
some kind of notable physical impairment count? Does a penguin who only
lived 1 year count? Does a penguin born in the Galápagos island but
subsequently transported to the Antarctic islands count?</p>
<p>No matter which research field you work in, you realize that
analogous questions appear when you try to define the “population” in
some study.</p>
<p>It is practically impossible to specify an exact criterion for
membership in a population under study. We may try to cover and delimit
as many factors as possible, but there may always appear a new one we
didn’t think of. There’s no “objective” specification of a population:
the specification depends on our research purpose – which often is not
precisely specified either. We must therefore always be prepared to
further specify the population of our study. In some cases we may even
need to modify our previous specification, and thus discard some data or
acquire new ones.</p>
<p>The intrinsic and unavoidable problem in the specification of a
population has important consequences for the way we observe and measure
the population and its variates. We now turn to this problem.</p>
</div>
<div id="sampling" class="section level4">
<h4>Sampling</h4>
<p>Our questions <strong>Q1</strong> and <strong>Q2</strong> could have
<em>exact statistical answers</em> if we had a <em>complete census</em>
of the penguin population; that is, if we went and “measured” the values
of the variates in each and every penguin of the population.</p>
<p>To make this point clearer, imagine a slightly different population.
Suppose we were interested only in all the penguins alive today on a
specific, very small island. The island turns out to have 17 penguins.
We check all of them. We find that 6 of these are females, and 11 are
males. It’s then a fact that 6/17${}$35.3% of penguins in this specific
population are females, and 64.7% are males. And if someone picked a
penguin from this population and asked us to guess its sex, we would
give a 35.3% probability to that penguin’s being female, and 64.7% to
its being male.<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a></p>
<p>But it is often impractical or impossible to take a complete census.
Think of the case were the whole population includes members or variates
that will only exist in the future. For this reason we proceed as
follows: we observe a <strong>sample</strong> of the whole population,
and from the study of this sample we try to infer the statistical
properties of the whole population. Our inferences are perforce
uncertain: we can’t be fully sure about the exact statistical properties
of the whole population. The essential point is that our uncertainty is
not just a matter of “I don’t know” versus “I know”:</p>
<ul>
<li>we can <em>quantify</em> our uncertainty;</li>
<li>we can <em>reduce</em> our uncertainty, sometimes to the point where
it becomes almost a certainty.</li>
</ul>
<p>Some trusted colleagues thus go to the three islands, and examine
samples of penguins there, sending back the values they observe. The way
they do the sampling would require a deep analysis and discussion, but
for the moment we put these aside.</p>
<p>How many samples should our colleagues collect? The short answer is:
as many as they can. The more samples we have, the more certain our
conclusions will be – all kinds of conclusions: for instance those that
may reveal the presence, as well as those that may reveal the absence,
of important associations.</p>
<p>One important feature of Bayesian inference is that <em>we don’t need
to choose beforehand when to stop sampling</em>. We can monitor the
results as more and more samples arrive, and stop as soon as the level
of certainty about the results is satisfactory. This is possible because
in Bayesian inference the <em>sample size affects only the
uncertainty</em> we have of the ground truth – association, effect, or
other hypothesis – but <em>it cannot affect the ground truth
itself</em>. If the truth is that “there is no effect”, then this truth
will come to light with more and more certainty as we increase the
sample size, and no amount of sampling will be able to change this
truth.</p>
</div>
</div>
<div id="first-analysis" class="section level2">
<h2>First analysis</h2>
<p>We receive the first 10 datapoints. They are stored in the file
<code>&#39;penguindata2007_10.csv&#39;</code>, respecting the <a href="#format">formatting rules required by
<strong><em>inferno</em></strong></a>. Here they are:</p>
<table style="width:100%;">
<colgroup>
<col width="4%" />
<col width="14%" />
<col width="14%" />
<col width="12%" />
<col width="12%" />
<col width="17%" />
<col width="14%" />
<col width="10%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">species</th>
<th align="left">island</th>
<th align="right">bill_len</th>
<th align="right">bill_dep</th>
<th align="right">flipper_len</th>
<th align="right">body_mass</th>
<th align="left">sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="left">Gentoo</td>
<td align="left">Biscoe</td>
<td align="right">47.8</td>
<td align="right">15.0</td>
<td align="right">215</td>
<td align="right">5650</td>
<td align="left">male</td>
</tr>
<tr class="even">
<td align="left">2</td>
<td align="left">Gentoo</td>
<td align="left">Biscoe</td>
<td align="right">45.4</td>
<td align="right">14.6</td>
<td align="right">211</td>
<td align="right">4800</td>
<td align="left">female</td>
</tr>
<tr class="odd">
<td align="left">3</td>
<td align="left">Chinstrap</td>
<td align="left">Dream</td>
<td align="right">52.0</td>
<td align="right">18.1</td>
<td align="right">201</td>
<td align="right">4050</td>
<td align="left">male</td>
</tr>
<tr class="even">
<td align="left">4</td>
<td align="left">Chinstrap</td>
<td align="left">Dream</td>
<td align="right">51.3</td>
<td align="right">18.2</td>
<td align="right">197</td>
<td align="right">3750</td>
<td align="left">male</td>
</tr>
<tr class="odd">
<td align="left">5</td>
<td align="left">Gentoo</td>
<td align="left">Biscoe</td>
<td align="right">42.9</td>
<td align="right">13.1</td>
<td align="right">215</td>
<td align="right">5000</td>
<td align="left">female</td>
</tr>
<tr class="even">
<td align="left">6</td>
<td align="left">Chinstrap</td>
<td align="left">Dream</td>
<td align="right">46.6</td>
<td align="right">17.8</td>
<td align="right">193</td>
<td align="right">3800</td>
<td align="left">female</td>
</tr>
<tr class="odd">
<td align="left">7</td>
<td align="left">Adelie</td>
<td align="left">Torgersen</td>
<td align="right">39.3</td>
<td align="right">20.6</td>
<td align="right">190</td>
<td align="right">3650</td>
<td align="left">male</td>
</tr>
<tr class="even">
<td align="left">8</td>
<td align="left">Adelie</td>
<td align="left">Torgersen</td>
<td align="right">38.6</td>
<td align="right">21.2</td>
<td align="right">191</td>
<td align="right">3800</td>
<td align="left">male</td>
</tr>
<tr class="odd">
<td align="left">9</td>
<td align="left">Chinstrap</td>
<td align="left">Dream</td>
<td align="right">51.7</td>
<td align="right">20.3</td>
<td align="right">194</td>
<td align="right">3775</td>
<td align="left">male</td>
</tr>
<tr class="even">
<td align="left">10</td>
<td align="left">Gentoo</td>
<td align="left">Biscoe</td>
<td align="right">48.7</td>
<td align="right">15.1</td>
<td align="right">222</td>
<td align="right">5350</td>
<td align="left">male</td>
</tr>
</tbody>
</table>
<p>With these datapoints we start our Bayesian nonparametric analysis
using <strong><em>inferno</em></strong>! Let’s load the package:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(inferno)</span></code></pre></div>
<p>There are two preliminary steps that we usually must follow to
perform an analysis:</p>
<ol style="list-style-type: decimal">
<li>Determine and prepare the <strong>metadata</strong> about the
variates of interest.</li>
<li>Perform the general inference about the whole population from the
sampled data; we shall call this <strong>learning</strong> from the
sample data.</li>
</ol>
<div id="metadata-preparation" class="section level4">
<h4>Metadata preparation</h4>
<p>Metadata are “data about the data”. They are provided in a CSV file
respecting the <a href="#format">formatting rules</a> and consist of
around eight pieces of information about the variates of our population.
Specifically:</p>
<dl>
<dt>Variate name (<code>name</code>)</dt>
<dd>
This is a character or string: name of a variate. Different variates
should obviously have different names.
</dd>
<dt>Variate type (<code>type</code>)</dt>
<dd>
<p><strong><em>inferno</em></strong> can handle three kinds of
variates:</p>
<ul>
<li><code>nominal</code>: it can take on a finite number of discrete
values, which do not have any natural ordering. Examples could be sex or
geographical location.</li>
<li><code>ordinal</code>: it can take on a finite number of discrete
values, which do have a natural ordering. They can be qualitative or
numeric. Examples could be the degree of satisfaction of a customer, or
the severity of a disease, or a <a href="https://www.britannica.com/topic/Likert-Scale">Likert
scale</a>.</li>
<li><code>continuous</code>: it can in principle take on an infinite
number of continuous values, although they can be discretized or
rounded. Examples could be age or weight.</li>
</ul>
<p>Many other types of variates exist. For example images and audio are
also types of variates; but <strong><em>inferno</em></strong> cannot
handle these complex types. A simple type of variate that
<strong><em>inferno</em></strong> cannot properly handle is the
<em>cyclic</em> one, such as time of day. There are no clear-cut
separations between different types of variates; thus it’s sometimes
difficult to assess the type.</p>
</dd>
<dt>Domain minimum and maximum (<code>domainmin</code>,
<code>domainmax</code>)</dt>
<dd>
<p>These are numbers, only defined for numeric-ordinal and continuous
variates. These metadata are the minimum and maximum value a variate can
take on in a given study. For example, in a study about health or
employment of adults, a variate <em>age</em> might have a minimum of
18 years. In a more general study involving people of all ages, the
minimum would be 0. In a health study where people of 90 years or more
are pooled together, the maximum would be 90 years. A minimum could also
be minus-infinity, and a maximum plus-infinity. Obviously the minimum
should be lower than the maximum.</p>
<p>Sometimes the minimum or maximum is not clear-cut. For instance,
there is no theoretical maximum on a person’s age, although we can
consider an age of 300 impossible to reach today. In such cases the
maximum or minimum can be taken to be plus or minus infinity.</p>
</dd>
<dt>Interval between values (<code>datastep</code>)</dt>
<dd>
This is a positive number, only defined for numeric-ordinal and
continuous variates. It is the separation between consecutive values.
For example, in a 1–5 Likert-scale variate, the interval is 1. For an
<em>age</em> variate expressed in and rounded to years, the interval
is 1. For a <em>length</em> variate expressed in centimetres and rounded
to millimetres, the interval is 0.1.
</dd>
<dt>Inclusion of inimum and maximum (<code>minincluded</code>,
<code>maxincluded</code>)</dt>
<dd>
These are logical or <code>yes</code>/<code>no</code>, only defined for
continuous variates. They tell whether the minimum and maximum of a
continuous variate are themselves possible values or not. Often such
extreme values have a special meaning, and this information is important
with censored or pooled data. For example, a <em>length</em> continuous
variate might be expressed in decimetres and reported without rounding,
but all lengths above 1 dm might be pooled together into the value “1 dm
or more”; in such case we must make sure to state that
<code>maxincluded</code> is <code>True</code> or <code>&quot;yes&quot;</code>.
</dd>
<dt>Values (<code>V1</code>, <code>V2</code>, …)</dt>
<dd>
These are characters or strings, only defined for nominal or non-numeric
ordinal variates. They are the values that a nominal or ordinal variate
can <em>in principle</em> take on in the population of interest. Note
that they are include not only the values that are observed in a sample,
but also those that could be observed in the rest of the population. A
nominal or ordinal variate must have at least two distinct values; it
wouldn’t make much sense to draw inferences about it otherwise.
</dd>
</dl>
<p>Why does <strong><em>inferno</em></strong> need metadata? Because it
performs <em>nonparametric</em> inference. In other words, it does not
assume the frequency distribution of the variates in the whole
population to have any specific class of shapes, such as a Gaussians. It
tries to extrapolate the frequency distribution of the whole population
from the sample data provided. We must therefore provide to it the same
general information that we have about the variates, and about any
artificial modifications performed on their values, such as rounding or
pooling.</p>
<p>In order to prepare the metadata file we can use the
<code>metadatatemplate()</code> helper function. This function reads the
sample data and prepares a preliminary file containing heuristic
<em>guesses</em> about the metadata, which we’ll have to check and
correct. The function motivates its guesses and warns about especially
uncertain ones:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>datafile <span class="ot">&lt;-</span> <span class="st">&#39;penguindata2007_10.csv&#39;</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>metadatafile <span class="ot">&lt;-</span> <span class="st">&#39;penguinmetadata.csv&#39;</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">metadatatemplate</span>(<span class="at">data =</span> datafile, <span class="at">file =</span> metadatafile)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co">#&gt; Analyzing 8 variates for 10 datapoints.</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; * &quot;species&quot; variate:</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt;   -  3 different  non-numeric  values detected:</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt;  &quot;Adelie&quot;, &quot;Chinstrap&quot;, &quot;Gentoo&quot; </span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt;   which do not seem to refer to an ordered scale.</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt;   Assuming variate to be NOMINAL.</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">#&gt; * &quot;island&quot; variate:</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt;   -  3 different  non-numeric  values detected:</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt;  &quot;Biscoe&quot;, &quot;Dream&quot;, &quot;Torgersen&quot; </span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt;   which do not seem to refer to an ordered scale.</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt;   Assuming variate to be NOMINAL.</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#&gt; * &quot;bill_len&quot; variate:</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="co">#&gt;   - Numeric values between 38.6 and 52 </span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="co">#&gt;   Assuming variate to be CONTINUOUS.</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="co">#&gt;   - Distance between datapoints is a multiple of 0.1 </span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="co">#&gt;   Assuming variate to be ROUNDED.</span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="co">#&gt;   - All values are positive</span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="co">#&gt;   Assuming &quot;domainmin&quot; to be 0</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a><span class="co">#&gt; * &quot;bill_dep&quot; variate:</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a><span class="co">#&gt;   - Numeric values between 13.1 and 21.2 </span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a><span class="co">#&gt;   Assuming variate to be CONTINUOUS.</span></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a><span class="co">#&gt;   - Distance between datapoints is a multiple of 0.1 </span></span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a><span class="co">#&gt;   Assuming variate to be ROUNDED.</span></span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a><span class="co">#&gt;   - All values are positive</span></span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a><span class="co">#&gt;   Assuming &quot;domainmin&quot; to be 0</span></span>
<span id="cb2-34"><a href="#cb2-34" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-35"><a href="#cb2-35" tabindex="-1"></a><span class="co">#&gt; * &quot;flipper_len&quot; variate:</span></span>
<span id="cb2-36"><a href="#cb2-36" tabindex="-1"></a><span class="co">#&gt;   - Only 9 different numeric values detected:</span></span>
<span id="cb2-37"><a href="#cb2-37" tabindex="-1"></a><span class="co">#&gt; from 190 to 222 in steps of 1 </span></span>
<span id="cb2-38"><a href="#cb2-38" tabindex="-1"></a><span class="co">#&gt;   Assuming variate to be ORDINAL.</span></span>
<span id="cb2-39"><a href="#cb2-39" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-40"><a href="#cb2-40" tabindex="-1"></a><span class="co">#&gt; * &quot;body_mass&quot; variate:</span></span>
<span id="cb2-41"><a href="#cb2-41" tabindex="-1"></a><span class="co">#&gt;   - Only 9 different numeric values detected:</span></span>
<span id="cb2-42"><a href="#cb2-42" tabindex="-1"></a><span class="co">#&gt; from 3650 to 5650 in steps of 25 </span></span>
<span id="cb2-43"><a href="#cb2-43" tabindex="-1"></a><span class="co">#&gt;   Assuming variate to be ORDINAL.</span></span>
<span id="cb2-44"><a href="#cb2-44" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-45"><a href="#cb2-45" tabindex="-1"></a><span class="co">#&gt; * &quot;sex&quot; variate:</span></span>
<span id="cb2-46"><a href="#cb2-46" tabindex="-1"></a><span class="co">#&gt;   -  2 different  values detected:</span></span>
<span id="cb2-47"><a href="#cb2-47" tabindex="-1"></a><span class="co">#&gt;  &quot;female&quot;, &quot;male&quot; </span></span>
<span id="cb2-48"><a href="#cb2-48" tabindex="-1"></a><span class="co">#&gt;   which do not seem to refer to an ordered scale.</span></span>
<span id="cb2-49"><a href="#cb2-49" tabindex="-1"></a><span class="co">#&gt;   Assuming variate to be NOMINAL.</span></span>
<span id="cb2-50"><a href="#cb2-50" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-51"><a href="#cb2-51" tabindex="-1"></a><span class="co">#&gt; * &quot;year&quot; variate:</span></span>
<span id="cb2-52"><a href="#cb2-52" tabindex="-1"></a><span class="co">#&gt;   Only one value present.</span></span>
<span id="cb2-53"><a href="#cb2-53" tabindex="-1"></a><span class="co">#&gt; =========</span></span>
<span id="cb2-54"><a href="#cb2-54" tabindex="-1"></a><span class="co">#&gt; WARNINGS - please make sure to check these variates in the metadata file:</span></span>
<span id="cb2-55"><a href="#cb2-55" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-56"><a href="#cb2-56" tabindex="-1"></a><span class="co">#&gt; * &quot;year&quot; variate does not have at least two distinct values.</span></span>
<span id="cb2-57"><a href="#cb2-57" tabindex="-1"></a><span class="co">#&gt; Discarded because non-informative.</span></span>
<span id="cb2-58"><a href="#cb2-58" tabindex="-1"></a><span class="co">#&gt; Please insert its characteristics by hand in the metadata file.</span></span>
<span id="cb2-59"><a href="#cb2-59" tabindex="-1"></a><span class="co">#&gt; =========</span></span>
<span id="cb2-60"><a href="#cb2-60" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-61"><a href="#cb2-61" tabindex="-1"></a><span class="co">#&gt; Saved proposal metadata file as &quot;penguinmetadata.csv&quot;</span></span></code></pre></div>
<p>Here is the preliminary metadata file created by this helper
function:</p>
<table style="width:100%;">
<colgroup>
<col width="9%" />
<col width="8%" />
<col width="7%" />
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="5%" />
<col width="7%" />
<col width="7%" />
<col width="2%" />
<col width="2%" />
<col width="2%" />
<col width="2%" />
<col width="2%" />
<col width="2%" />
<col width="3%" />
<col width="3%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">name</th>
<th align="left">type</th>
<th align="right">domainmin</th>
<th align="right">domainmax</th>
<th align="right">datastep</th>
<th align="right">minincluded</th>
<th align="right">maxincluded</th>
<th align="left">V1</th>
<th align="left">V2</th>
<th align="left">V3</th>
<th align="right">V4</th>
<th align="right">V5</th>
<th align="right">V6</th>
<th align="right">V7</th>
<th align="right">V8</th>
<th align="right">V9</th>
<th align="right">V10</th>
<th align="right">V11</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">species</td>
<td align="left">nominal</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="left">Adelie</td>
<td align="left">Chinstrap</td>
<td align="left">Gentoo</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td align="left">island</td>
<td align="left">nominal</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="left">Biscoe</td>
<td align="left">Dream</td>
<td align="left">Torgersen</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td align="left">bill_len</td>
<td align="left">continuous</td>
<td align="right">0</td>
<td align="right"></td>
<td align="right">0.1</td>
<td align="right"></td>
<td align="right"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td align="left">bill_dep</td>
<td align="left">continuous</td>
<td align="right">0</td>
<td align="right"></td>
<td align="right">0.1</td>
<td align="right"></td>
<td align="right"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td align="left">flipper_len</td>
<td align="left">ordinal</td>
<td align="right">190</td>
<td align="right">222</td>
<td align="right">1.0</td>
<td align="right"></td>
<td align="right"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td align="left">body_mass</td>
<td align="left">ordinal</td>
<td align="right">3650</td>
<td align="right">5650</td>
<td align="right">25.0</td>
<td align="right"></td>
<td align="right"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td align="left">sex</td>
<td align="left">nominal</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="left">female</td>
<td align="left">male</td>
<td align="left"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
</tbody>
</table>
<p>We see that <code>metadatatemplate()</code> guessed correctly about
the <code>species</code>, <code>island</code>, and <code>sex</code>
variates, assigning them a <code>nominal</code> type and listing all
possible values that can occur in the whole population. Note that if we
knew of an additional species present in the whole population but absent
from the sample, we ought to add it under the <strong>V3</strong>
column.</p>
<p>The guesses about the <code>bill_len</code> and <code>bill_dep</code>
variates are also correct: they are continuous, rounded to the nearest
tenth of millimetre, and cannot be less than zero. The absence of
<strong>domainmax</strong> values means that the theoretical maximum is
plus infinity. We could replace this maximum with something more
realistic, for instance <code>1000</code> mm or less, but typically
these changes do not lead to relevant differences in our inferences. The
absence of <strong>minincluded</strong> means that the minimum,
<code>0</code>, is not a possible value; similarly for
<strong>maxincluded</strong>.</p>
<p>The guess about the <code>flipper_len</code> variate is not correct.
This variate is similar to <code>bill_len</code> but rounded to the
nearest integer. Similarly for the wrongly guessed
<code>body_mass</code> variate, which is rounded in steps of
<code>25</code>.</p>
<p>We must open and edit the metadata file
<code>&#39;penguindata2007_10.csv&#39;</code> with our favourite editor and
correct and complete the guesses of the helper function. In this case we
end up with the following corrected metadata file:</p>
<table style="width:100%;">
<colgroup>
<col width="9%" />
<col width="8%" />
<col width="7%" />
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="5%" />
<col width="7%" />
<col width="7%" />
<col width="2%" />
<col width="2%" />
<col width="2%" />
<col width="2%" />
<col width="2%" />
<col width="2%" />
<col width="3%" />
<col width="3%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">name</th>
<th align="left">type</th>
<th align="right">domainmin</th>
<th align="right">domainmax</th>
<th align="right">datastep</th>
<th align="right">minincluded</th>
<th align="right">maxincluded</th>
<th align="left">V1</th>
<th align="left">V2</th>
<th align="left">V3</th>
<th align="right">V4</th>
<th align="right">V5</th>
<th align="right">V6</th>
<th align="right">V7</th>
<th align="right">V8</th>
<th align="right">V9</th>
<th align="right">V10</th>
<th align="right">V11</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">species</td>
<td align="left">nominal</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="left">Adelie</td>
<td align="left">Chinstrap</td>
<td align="left">Gentoo</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td align="left">island</td>
<td align="left">nominal</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="left">Biscoe</td>
<td align="left">Dream</td>
<td align="left">Torgersen</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td align="left">bill_len</td>
<td align="left">continuous</td>
<td align="right">0</td>
<td align="right"></td>
<td align="right">0.1</td>
<td align="right"></td>
<td align="right"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td align="left">bill_dep</td>
<td align="left">continuous</td>
<td align="right">0</td>
<td align="right"></td>
<td align="right">0.1</td>
<td align="right"></td>
<td align="right"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td align="left">flipper_len</td>
<td align="left">continuous</td>
<td align="right">0</td>
<td align="right"></td>
<td align="right">1.0</td>
<td align="right"></td>
<td align="right"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td align="left">body_mass</td>
<td align="left">continuous</td>
<td align="right">0</td>
<td align="right"></td>
<td align="right">25.0</td>
<td align="right"></td>
<td align="right"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td align="left">sex</td>
<td align="left">nominal</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="left">female</td>
<td align="left">male</td>
<td align="left"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
</tbody>
</table>
<p>Later on we shall add the variate <code>year</code> to the metadata
file.</p>
</div>
<div id="learning-and-extrapolating-from-the-sample-data" class="section level4">
<h4>Learning and extrapolating from the sample data</h4>
<p>Now comes the essential part of our analysis: from the sample data
and the metadata information,<strong><em>inferno</em></strong> will draw
inferences about the <em>whole</em> population of penguin that we have
specified.</p>
<p><br />
<br />
</p>
<hr />
</div>
</div>
<div id="format" class="section level2">
<h2>Format for data and metadata files and their contents</h2>
<p>The functions of the <strong><em>inferno</em></strong> package accept
CSV files formatted as follows:</p>
<ul>
<li>Decimal values should be separated by a <em>dot</em>; no comma
should be used to separate thousands etc. Example:
<code>86342.75</code> .</li>
<li>Character and names should be quoted in single or double quotes.
Example: <code>&quot;female&quot;</code>.</li>
<li>Values should be separated by <em>commas</em>, not by tabs or
semicolons.</li>
<li>Missing values should be simply <em>empty</em>, not denoted by “NA”,
“missing”, “-”, or similar.</li>
</ul>
<p>Names of variates and character variate values should be strings
conforming to <a href="https://cran.r-project.org/doc/FAQ/R-FAQ.html#What-are-valid-names_003f">R’s
rules</a>.</p>
<p><em>(TO BE CONTINUED)</em></p>
<!-- 

## Population inferences

A very common type of inference is the following. We have a large or potentially infinite group of entities, each having a set of characteristics. Having checked the characteristics of several entities from this group, we want to make guesses about the characteristics of new entities from the same group.

The description just given is very generic or even vague. Indeed it can be applied to a huge variety of situations.

- The 'entities' could be: objects, like electronic components; or people, or animals, or plants, or events of some kind.
- The characteristics could be: the material the object is made of, and the result of some test made on it; or the age, blood-test results, and disease condition of the person; or the species and body mass of the animal; or the length of the petals of the flower.
- The group could be: electronic components coming out of a particular production line; or people of a given nationality and with given symptoms; or animals from a particular island; or flowers of a specific genus or family.

The possibilities are endless.

Also the kinds of guesses that we want to make can be very diverse. We might want to guess all characteristics of a new entity; or, having checked *some* characteristics of a new entity, we want to guess the ones that we haven't or can't check. Important examples of this kind of inference appear in medicine. For example we may have a group defined as follows:

- *Group*: people from a given nationality, suffering from one of two possible diseases.

And we may consider these characteristics:

- *Characteristics*: age; sex; weight; presence or absence of a particular genetic factor; symptoms from a specific set of possible ones; results from clinical tests taken at different times; kind of disease.

We observed as many as possible of these characteristics in a sample of people from this group. Now a new person from the same group appears in front of us. We check this person's age, sex, weight, symptoms. We need to guess which of the two diseases affects this person.

\
The kind of inference summarily described above has one important aspect. Suppose you have collected a sample from your group of interest, and you want to use this sample to make guesses about new entities from the same group. If someone exchanged one entity in your sample with another one *unsystematically* chosen from the group, then you wouldn't protest. After all, you still have the same number of samples from the same group. This aspect is called **exchangeability**: we say that this kind of inference is exchangeable. There are kinds of inference for which this aspect is not true. For example, suppose you're given some stock data from four consecutive days, and you want to make guesses about the next day. If someone replaced any of your four datapoints with a datapoint an unsystematically chosen other day of the year, then you would protest: the time order of the datapoints matter. This is an example of non-exchangeable inference.

The inferences for which *inferno* is designed are exchangeable ones. We shall also call them **population inferences**.

\
Having discussed these simple examples, let's agree on some more standard terminology. Let's call:

- **Population**: what we've called 'group'.
- **Unit**: what we've called 'entity' -- the object, person, animal, etc.
- **Variate**: what we've called 'characteristic'.


## Probability and population frequencies

We have been speaking about "making guesses"; but what does this mean?

When a unit is chosen unsystematically from a population, it's only in rare situations that we may be sure about the variates of this unit before checking them. Suppose for instance that the population of interest is 'all adults from a given country'; and the variates of interest are `sex`, `weight`, `height`. If a person is chosen from this population, we can't be sure beforehand of how tall this person will be. We can exclude values like 4 m or 20 cm, but we'll be uncertain about many other possible values. Even if someone tells us the sex and weight of this person, we'll still be in doubt, but maybe we can consider some values more probable than others.

That's the keyword: **probability**. Although we are unsure about a variate of a unit, we can still find some values more probable than others: we may consider it more probable that the person is 160 cm tall than 180 cm tall; or in a clinical inference we may consider it more probable that the patient has a particular virus than not.

Probability is our *degree of belief* about the value of the unknown variate.

Note that probability is not a physical property of the unit. For instance, the person in question may be exactly 158 cm tall; the probability of being 180 cm tall is not something we can "measure" from the person. Also, if we found out some other variate of the person -- say we knew the weight, and now we know also the sex -- then the probability we assign to 180 cm may increase or decrease; but the person is exactly the same as before. Probability is not a property of the population either. For instance, for one person from a population we may think 150 cm to be the most probable height, but for another person from the same population we may think 180 cm to be the most probable height instead.

Probability expresses the *information* we have about the variate of a unit. This is why another researcher may have a different probability about the same unit: because they may have different information about that unit.

\
There is a situation in which we would all agree about the probability about a variate of a unit: when we know the **frequencies** for that variate in the population. Suppose for instance that the population of interest is that of penguins who lived in particular locations in some particular years (penguins from other locations or other times are not part of this population). The variates are the penguins' `species` and the `island` they lived in.

You're told that, as a matter of fact, 43.7% of penguins from the *whole* population are of species *Adelie*, 20.0% of species *Chinstrap*, and 36.3% of species *Gentoo*. Now a penguin from that population is brought in front of you, but you can't see any of its characteristics. What's your degree of belief that this penguin is of species *Adelie*, or *Chinstrap*, or *Gentoo*? We'd all agree, given the frequency information about this population, to assign the probabilities of 43.7%, 20.0%, and 36.3% to the three possibilities, for this penguin. We write this as follows:
$$
\begin{aligned}
&\mathrm{Pr}(\texttt{species}\!=\!\mathit{Adelie} \mid 
\textsf{population frequency}) = 0.437
\\
&\mathrm{Pr}(\texttt{species}\!=\!\mathit{Chinstrap} \mid 
\textsf{population frequency}) = 0.200
\\
&\mathrm{Pr}(\texttt{species}\!=\!\mathit{Gentoo} \mid 
\textsf{population frequency}) = 0.363
\end{aligned}
$$
On the left side of the '$\mid$' bar we write what we're guessing or are unsure about. On the right side, we write the information that led to our probability assignment; in this case, the information about the full population. When the information behind a probability is understood, the '$\mid$' bar and the right side are usually omitted.

In this case, the highest probability is for the *Adelie* species, but the other two possibilities cannot be excluded. The numerical values of these probabilities are extremely important, because they determine any kind of **decision** we may have to make about our unit. This is especially true in medical decision-making, where probabilities, combined with **utilities**, determine which is the best choice that a clinician can make. Medical and clinical decision-making, and the role of probabilities in them, are discussed for instance in the texts by Sox & al.: [*Medical Decision Making* (2024)](https://doi.org/10.1002/9781119627876), by Hunink & al.: [*Decision Making in Health and Medicine* (2014)](https://doi.org/10.1017/CBO9781139506779), or by Weinstein & Fineberg [*Clinical Decision Analysis* (1980)](https://archive.org/details/clinicaldecision0000unse_e3n8).

The example with the penguin population can be analysed further. The penguins in this population come from three possible `island`s: *Biscoe*, *Dream*, and *Torgersen*. Therefore a penguin can be of one of three species and from one of three islands, for a total of 3 × 3 = 9 possibilities. You're told that, as a matter of fact, the frequencies of these nine combinations in the whole population are as follows:

|             |          |             |          |
|:-----------:|:--------:|:-----------:|:--------:|
|             | *Adelie* | *Chinstrap* | *Gentoo* |
| *Biscoe*    | 12.6%    | 1.8%        | 34.0%    |
| *Dream*     | 17.4%    | 16.5%       | 1.4%     |
| *Torgersen* | 13.8%    | 1.6%        | 0.9%     |

Then we'd all agree that the probability that the penguin brought to us is of `species` *Gentoo* and from `island` *Biscoe* would be 34.0%:
$$
\mathrm{Pr}(
\texttt{species}\!=\!\mathit{Gentoo} ,
\texttt{island}\!=\!\mathit{Biscoe} 
\mid 
\textsf{population frequency}) = 0.340
$$

\
***inferno*** allows you to calculate probabilities of this kind, for any set of variates of your choice.

## Learning from known variates

But there's even more interesting information in the population frequencies above. Let's focus on *Biscoe* island. With a quick sum we see that 48.4% of the whole population comes from *Biscoe* island (thus we'd assign a probability of 0.484 that our penguin comes from that island). Dividing the frequencies above, row-wise, by the frequencies of the respective islands (the sums of each row), we can find the frequency of each species *for each particular island*:

|                  |          |             |          |
|:----------------:|:--------:|:-----------:|:--------:|
|                  | *Adelie* | *Chinstrap* | *Gentoo* |
| from *Biscoe*    | 26.03%   | 3.72%       | 70.25%   |
| from *Dream*     | 49.29%   | 46.74%      | 3.97%    |
| from *Torgersen* | 84.66%   | 9.82%       | 5.52%    |

 We call these **conditional frequencies**, and we call the group of penguins that come from `island` *Biscoe*a **subpopulation** of the whole population. The table above reports the frequencies of the three `species` in each subpopulation.

Thus we also know, for instance, that 70.25% of penguins in the subpopulation from `island` *Biscoe* are of `species` *Gentoo*. This species is the majority in that subpopulation -- contrast this with the majority in the whole population, which we saw was *Adelie*. Indeed, your most probable guess about the `species` of the penguin in front of you was *Adelie*, with 0.437 probability. 

But suppose now someone tells you that this penguin comes from `island` *Biscoe* (so this variate is now known to you). Given this new piece of information, which probabilities do you assign to the `species` of this penguin? Obviously 0.2603 for *Adelie*, 0.0372 for *Chinstrap*, and 0.7025 for *Gentoo*. We write this as follows:
$$
\begin{aligned}
&\mathrm{Pr}(\texttt{species}\!=\!\mathit{Adelie} \mid 
\texttt{island}\!=\!\mathit{Biscoe} ,
\textsf{population frequency}) = 0.2603
\\
&\mathrm{Pr}(\texttt{species}\!=\!\mathit{Chinstrap} \mid 
\texttt{island}\!=\!\mathit{Biscoe} ,
\textsf{population frequency}) = 0.0372
\\
&\mathrm{Pr}(\texttt{species}\!=\!\mathit{Gentoo} \mid 
\texttt{island}\!=\!\mathit{Biscoe} ,
\textsf{population frequency}) = 0.7025
\end{aligned}
$$
The right side of the '$\mid$' bar now reports the extra information that the penguin comes from `island` *Biscoe*. We say that our probability has been *updated*, and we call it a **conditional probability**.

Learning about the penguin's `island` not only made you change the highest probability assignment from *Aelie* to *Gentoo*, but it also increased the value of the highest probability. Without knowing the penguin's `island`, *Adelie* had slightly less than 50% probability; it was as likely as not. After learning the `island` variate, *Gentoo* gets more than 70% probability.

An analogous discussion can be made for a continuous variate. Let's for instance take the `body_mass` of the penguins, discretized in steps of 0.025 kg. Suppose we knew that the histogram of body masses in the whole population were as follows:

<img src="prob-body_mass.jpg" width="75%" style="display: block; margin: auto;" />

Then, if you had to guess the body mass of the penguin brought to you, you'd give the value 4 kg (plus or minus 0.0125 kg) a probability of around 1%, and so on for other possible values; the most probable value being 3.6 kg at 1.4% probability.

Suppose you're now given also the histograms for the subpopulations from the three islands:

<img src="prob-body_mass-given_island.jpg" width="75%" style="display: block; margin: auto;" />

Upon learning that your penguin is from *Biscoe*, your degree of belief would change: the most probable value would be 4.8 kg at 1.3% probability. Note how the histograms for `body_mass` are different in the subpopulations of the three islands. In *Torgersen*, for instance, it's more probable to find penguins between 3 kg and 4 kg than between 5 kg and 6 kg; whereas in *Biscoe* the opposite is true.

\medskip

This is exactly the kind of learning situation that takes place in medicine and in clinical inferences. A clinician searches for symptoms because they may change and increase the probabilities of different diseases or health conditions. The reason why clinicians research particular *subpopulations* -- particular demographics, or genetic factors, etc. -- is that within these subgroups the probability that a medical condition exist or will occur can be drastically higher. In turn, this leads the clinicians to understand better what can be the biological relationships between the health condition and those factors.

\
***inferno*** allows you to calculate the updated probabilities of any set of variates, conditional on any other set.


## Uncertainty about whole populations. Population samples

Inferences are therefore quite clear if we know the frequencies of the variates for the *whole* population. Our main problem is that we usually *don't know those whole-population frequencies*. For example, do you know the *exact* number of people, among those born in your country at any time and are alive today, who are exactly between 165 cm and 170 cm tall today?

For our penguin example, let's consider again the `body_mass` variate. We actually don't know what the histogram of this variate over the whole population looks like. It could have one or several peaks, a symmetric or non-symmetric shape, shoulders, an so on. Here are seven possibilities:

<img src="penguin_priors.jpg" width="75%" style="display: block; margin: auto;" />

\
It's often impractical or impossible to measure the frequencies of a variate in a whole population. We must therefore *guess* them.

In order to guess them, we usually examine a **sample** from the population of interest. This sample is chosen in an unsystematic way, so that its statistics and its **sample frequencies** are not affected by peculiar choices, otherwise we would have a *biased* sample (imagine for instance choosing a sample of *male* penguins only, whereas the penguin population we're interested in has both males and females).

But the frequencies of variates in a sample from a population are typically different from those in the whole population. The difference is the larger, the smaller the sample. For example, if we chose five penguins from our penguin population in an unsystematic way, it could happen that we got five penguins all of species *Gentoo*. Then the variate `species` would have frequencies of 100% for *Gentoo* and 0% for *Adelie* and *Chinstrap*. But of course this would not mean that only the species *Gentoo* occurs in the whole population. In order to reflect the frequencies in the whole population, the size of a sample needs to be larger, the larger are:

- the numbers of variates considered;
- the possible values of each variate.

Therefore, to guess the frequencies in the whole population we can examine a sample from it -- but we cannot fully rely on the sample.

\medskip

The traditional way of facing this uncertainty was to *assume* that the whole population has frequencies with particular features; most commonly, they were assumed to be *Gaussian*. The variate values observed in the sample were used to fit the free parameters of the particular assumed distribution, like the means, variances, covariances of the Gaussian.

Consider how drastic such an approach is. It can also be misleading, if one is not aware of what is being done. For instance, errors can be reported about the fit of the free parameters, and if they are small they can give a false sense of precise inference. But the very *assumption* has errors, which may be quite large; and such errors are rarely reported.

One could object that the assumption, say of gaussianity, behind such methods is not completely arbitrary: one can get the idea of whether it's correct or not by looking at the sample. But this objection is self-contradictory: as discussed, our problem is that the frequencies in the sample are not a faithful reflection of those in the whole population. We're making assumption because the sample is not reliable, so how can we rely on the sample to judge those assumptions?

It must be pointed out that even methods or measures that do not explicitly refer to peculiar assumptions about frequencies, *do* actually rely on such hidden assumptions. Pearson's measure of correlation, for instance, actually relies on the assumption that the joint frequency distribution for the whole population is Gaussian.
 add ref 

## Bayesian nonparametric population inference

The method based on the assumption of peculiar frequency distribution was a necessity in the past, because there was no other computationally feasible way of facing the uncertainty about the whole population. But today the situation is different.

Instead of making peculiar assumptions out of computational desperation, today we can explicitly recognize our uncertainty about the frequencies in the whole population, and therefore account for the errors that can come from this uncertainty. This is what Bayesian nonparametric population inference does. The term 'nonparametric' means that no peculiar assumptions are made about frequency distribution having simple shapes (and therefore expressible with few parameters).

Since we must *guess* what the frequencies are in the whole population, we proceed by assigning *probabilities* to all possible such frequency distributions. Intuitively, a candidate frequency distribution is more probable:

- the more it fits the frequencies measured in the sample;
- the more it looks "natural".

An example of unnatural or strange probability distribution, for our `body_mass` variate, could be the following:

<img src="penguin_prior_strange.jpg" width="75%" style="display: block; margin: auto;" />

It seems quite unnatural based on our experience with histograms over large natural populations. If a sample from a penguin population showed this frequency distribution of `body_mass` values, you would probably check whether any errors were made during the sampling.

The core of the ***inferno*** package is the calculation of the probabilities for the possible frequencies of the whole population, and the selection of a large sample of the most probable ones. This calculation and sampling are realized by the `learn()` function.

## Back to guessing about the next unit






There are situations in which it is not possible to choose the sample in an unsystematic way. Yet such a biased sample can still be used for population inference, as long as we know what its bias may be. ***inferno*** allows you to make this kind of bias corrections as well.

But the frequencies of variates in a sample from a population are typically different from those in the whole population. The difference is the larger, the smaller the sample. There are even extreme situations in which some statistics from a sample, like the mean, can be way off those of the whole population, even if the sample is large (the statistics become reliable only when the sample is almost as large as the full population). The plot below shows an example. The variate is a real number that can take on positive or negative values, in a population of 100 000 000 units; the mean value of the variate in the whole population is $-1.14$. The plot shows the values of the mean obtained from unsystematically chosen samples of various sizes. Even in a sample of 30 000 000 units we may observe a mean of $-3$. The *median* of a sample is much more reliable.

<img src="sample_mean.jpg" width="67%" style="display: block; margin: auto;" />

The point of this example is only to make clear that although a sample from a population gives us some information about the whole population, we are still uncertain about the whole-population frequencies.


<!-- ![](sample_mean.jpg){#id .class width=50%}

*(TO BE CONTINUED)*


----

(OLD TEXT)



We store the name of the datafile

``` r
dataset <- penguins
```
Here are the values for two subjects:


|    |species   |island    | bill_len| bill_dep| flipper_len| body_mass|sex    | year|
|:---|:---------|:---------|--------:|--------:|-----------:|---------:|:------|----:|
|225 |Gentoo    |Biscoe    |     48.2|     15.6|         221|      5100|male   | 2008|
|127 |Adelie    |Torgersen |     38.8|     17.6|         191|      3275|female | 2009|
|315 |Chinstrap |Dream     |     46.9|     16.6|         192|      2700|female | 2008|
|189 |Gentoo    |Biscoe    |     42.6|     13.7|         213|      4950|female | 2008|
|271 |Gentoo    |Biscoe    |     47.2|     13.7|         214|      4925|female | 2009|



### "Natural" vs controlled or "biased" variates

## Metadata

Let's load the package:

``` r
library(inferno)
```


```{.r .fold-hide}
## metadatatemplate(data = datafile, file = 'temp_metadata')
```


``` r
metadatafile <- 'meta_toydata.csv'
```
The metadata are as follows; `NA` indicate empty fields:


|name                                    |type       | datastep| domainmin| domainmax|minincluded |maxincluded |V1     |V2      |
|:---------------------------------------|:----------|--------:|---------:|---------:|:-----------|:-----------|:------|:-------|
|TreatmentGroup                          |nominal    |         |          |          |            |            |NR     |Placebo |
|Sex                                     |nominal    |         |          |          |            |            |Female |Male    |
|Age                                     |continuous |        1|         0|          |            |            |       |        |
|Anamnestic.Loss.of.smell                |nominal    |         |          |          |            |            |No     |Yes     |
|History.of.REM.Sleep.Behaviour.Disorder |nominal    |         |          |          |            |            |No     |Yes     |
|MDS.UPDRS..Subsection.III.V1            |ordinal    |        1|         0|       132|yes         |yes         |       |        |
|diff.MDS.UPRS.III                       |ordinal    |        1|      -132|       132|yes         |yes         |       |        |



*(More to be written here!)*


## Learning

$$
\mathrm{Pr}(Y = y \:\vert\: X = x, \mathrm{data})
$$
As the notation above indicates, *these probabilities also depend on the $\mathrm{data}$ already observed*. They are usually called "posterior probabilities".

We need to prepare the software to perform calculations of posterior probabilities given the observed data. In machine learning an analogous process is called "learning". For this reason the function that achieves this is called `learn()`. It requires at least three arguments:

- `data`, which can be given as a path to the `csv` file containing the data
- `metadata`, which can also be given as a path to the metadata file
- `outputdir`: the name of the directory where the output should be saved.

It may be useful to specify two more arguments:

- `seed`: the seed for the random-number generator, to ensure reproducibility
- `parallel`: the number of CPUs to use for the computation

Alternatively you can set the seed with `set.seed()`, and start a set of parallel workers with the `parallel::makeCluster()` and `doParallel::registerDoParallel()` commands.

The "learning" computation can take tens of minutes, or hours, or even days depending on the number of variates and data in your inference problem. The `learn()` function outputs various messages on how the computation is proceeding. As an example:


``` r
learnt <- learn(
    data = datafile,
    metadata = metadatafile,
    outputdir = 'parkinson_computations',
    seed = 16,
    parallel = 12)

#> Registered doParallelSNOW with 12 workers
#>
#> Using 30 datapoints
#> Calculating auxiliary metadata
#>
#> **************************
#> Saving output in directory
#> parkinson_computations
#> **************************
#> Starting Monte Carlo sampling of 3600 samples by 60 chains
#> in a space of 703 (effectively 6657) dimensions.
#> Using 12 cores: 60 samples per chain, 5 chains per core.
#> Core logs are being saved in individual files.
#>
#> C-compiling samplers appropriate to the variates (package Nimble)
#> this can take tens of minutes with many data or variates.
#> Please wait...
```





## Drawing inferences



*(TO BE CONTINUED)*

 -->
</div>
<div class="footnotes footnotes-end-of-document">
<hr />
<ol>
<li id="fn1"><p>Note the difference between probability and frequency.
If an expert colleague tells us “I was able to take a quick look at that
penguin, and it looks female to me, though I’m not completely sure”,
then our <em>probability</em> that this specific penguin is female would
get higher than 50%. If we observe the penguin and see that it’s female,
our probability for female would become 100%. Yet the <em>frequency</em>
of females in the population is still 35.3%.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
